File: Java/src/_0072_Edit_Distance/Solution.java
package _0072_Edit_Distance;

/**
 * Date: 2021/4/25 22:05
 */
public class Solution {

    public int minDistance(String word1, String word2) {

        int n = word1.length(), m = word2.length();
        char[] a = (" " + word1).toCharArray(), b = (" " + word2).toCharArray();

        int[][] f = new int[n + 1][m + 1];

        for (int i = 0; i <= m; i++) f[0][i] = i;
        for (int i = 0; i <= n; i++) f[i][0] = i;

        for (int i = 1; i <= n; i++)
            for (int j = 1; j <= m; j++) {
                f[i][j] = Math.min(f[i - 1][j], f[i][j - 1]) + 1;
                if (a[i] == b[j]) f[i][j] = Math.min(f[i][j], f[i - 1][j - 1]);
                else f[i][j] = Math.min(f[i][j], f[i - 1][j - 1] + 1);
            }
        return f[n][m];
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).minDistance("horse", "ros"));  // 3
    }
}

File: Python/_0072_Edit_Distance/Solution.py
# coding=utf-8
# Date: 2021/5/26 16:54


class Solution:
    def minDistance(self, a: str, b: str) -> int:
        n = len(a); m = len(b)
        a = " " + a; b = " " + b
        f = [[0 for _ in range(m + 1)] for _ in range(n + 1)]
        for j in range(m + 1):
            f[0][j] = j
        for i in range(n + 1):
            f[i][0] = i
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                f[i][j] = min(f[i - 1][j] + 1, f[i][j - 1] + 1)
                if a[i] == b[j]:
                    f[i][j] = min(f[i][j], f[i - 1][j - 1])
                else:
                    f[i][j] = min(f[i][j], f[i - 1][j - 1] + 1)
        return f[n][m]


if __name__ == "__main__":
    print(Solution().minDistance("horse", "ros"))  # 3
